### ISSUE TITLE
Enhancement: [no-magic-numbers] new option for ignoring generic types

### ISSUE THREAD

**Issue by DerHerrGammler** [typescript-eslint/typescript-eslint#6044]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-magic-numbers/

### Description

In my project i work a lot with generic types and i also set some default in this generics sometimes. So i think it would be a good option to add a `ignoreTypeGenerics` or somthing similar to ignore those cases.

### Fail

```typescript
type Enumerate<N extends number, Acc extends number[] = []> = Acc["length"] extends N
    ? Acc[number]
    : Enumerate<N, [...Acc, Acc["length"]]>;

export type IntRange<F extends number, T extends number> = Exclude<Enumerate<T>, Enumerate<F>>;

export type TColumnRange = IntRange<1, 13>; // fail with ignore = false
```

### Pass

```typescript
type Enumerate<N extends number, Acc extends number[] = []> = Acc["length"] extends N
    ? Acc[number]
    : Enumerate<N, [...Acc, Acc["length"]]>;

export type IntRange<F extends number, T extends number> = Exclude<Enumerate<T>, Enumerate<F>>;

export type TColumnRange = IntRange<1, 13>; // pass with ignore = true
```

**Comment by bradzacher** [comment]
Personally I don't see any difference here between these two cases conceptually:

```ts
const bar = IntRange(1, 13) ; 
type foo = IntRange<1, 13>;
```

The base rule does not provide an option to ignore the former, so IMO the extension rule should not provide an option to ignore the latter either.

**Comment by DerHerrGammler** [comment]
In the first case of your examle you are generating something into a variable.  
My example ist setting up a type that has some numbers in it that are valid for some variables. I think there is definitly a difference between this code examples.

```typescript
type foo = IntRange<1, 13>;
// is same to
type bar = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;


const baz = IntRange(1, 13);
console.log(baz);
// would end in >>> "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11. 12]"
```

First example is a type secound example is executed code. Definetly not the same and the `type bar` is ignorable by adding option `ignoreNumericLiteralTypes`

**Comment by bradzacher** [comment]
The point I was trying to get at is that a generic type that accepts two numbers as arguments is akin to a function cal that accepts two numbers as arguments.
They're both "reusable procedures that generate some output" with the difference being that one operates in the value space and the other operates in the type space.

If the rule has no way to ignore the the function call argument case and it so (IMO) it should have no way to ignore the generic type argument case.

`ignoreNumericLiteralTypes` lets you write the same code in type-space as you would in value-space:

```ts
const rangeStart = 1;
const rangeEnd = 13;
const foo = IntRange(rangeStart, rangeEnd);

type RangeStart = 1;
type RangeEnd = 13;
type Foo = IntRange<RangeStart, RangeEnd>; 
```
[playground](https://typescript-eslint.io/play#ts=4.8.4&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXy1QwCUpUBzEACigC55VkBbAIxBgBp4X7HX2AlPTIBPANwAoCWDwBnDPBhlKAZQywFAXngBGSTNTzFykAFFUweNp0BmfXIWIcOKwSKkK1JZ7Uau3ynNgAUkpDBEABwQASXcTAB4AQS4AIQA+V1FQ8Kj4D1V1GC1dSRyEfLMLV1tSyIQAMWdXWJIEit8irgqgtMkgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFsBDAc0plaTWNwBGiaKgwBtcNhxToHaJAA0CxVkXZIlfp2iIAcmKlCAMpXxTe8ACokU6MPjiIt2AL5eAugq+3kA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

Thus there's consistency and parity in the rule and the codebase.

For the above reason, I am going to reject this request.
